### ISSUE TITLE
Pytorch Geometric support for other Neighborhood Sampling methods

### ISSUE THREAD

**Issue by vymao** [pyg-team/pytorch_geometric#1787]
## ❓ Questions & Help

Just wondering if Pytorch Geometric supports other Neighborhood Sampling methods other than the one described in the original GraphSAGE paper. There seem to have been several other ones developed, some of which have been implemented through DGL (https://www.dgl.ai/blog/2019/06/13/giant.html).

**Comment by rusty1s** [comment]
We currently provide the following techniques for operating on large scale graphs:
* [NeighborSampler](https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.NeighborSampler) (GraphSAGE)
* [ClusterLoader](https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.ClusterLoader) (Cluster-GCN)
* [GraphSAINTSampler](https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.GraphSAINTSampler) (GraphSAINT)
* [SIGN](https://pytorch-geometric.readthedocs.io/en/latest/modules/transforms.html#torch_geometric.transforms.SIGN)
We will likely not implement layer-wise sampling though.

**Comment by vymao** [comment]
Thanks. I am confused about how GraphSAINT has been implemented. To my understanding, the edge and node weights should be used to calculate the aggregation and loss normalization. In the example `graph_saint.py`, there is this section for each batch: 
```
        if args.use_normalization:
            edge_weight = data.edge_norm * data.edge_weight
            out = model(data.x, data.edge_index, edge_weight)
            loss = F.nll_loss(out, data.y, reduction='none')
            loss = (loss * data.node_norm)[data.train_mask].sum()
        else:
            out = model(data.x, data.edge_index)
            loss = F.nll_loss(out[data.train_mask], data.y[data.train_mask])
```
Is there an explanation in how this relates to the paper derivation? I do not see how it translates equivalently.

**Comment by rusty1s** [comment]
You are right that GraphSAINT uses non-optional normalization. I just made it optional for the sake of simplicity (as it complicates the implementation).
These lines refer to Eq. (2) and Eq. (3) in the paper:
```python
edge_weight = data.edge_norm * data.edge_weight
```
normalizes the adjacency matrix (Eq. (2)), and
```python
loss = (loss * data.node_norm)
```
normalizes the loss (Eq. (3)).

**Comment by vymao** [comment]
So are the `edge_norm` and `node_norm` values the inverse of the normalization factors alpha and lambda in the paper (which are divided by for normalization), if they are being multiplied here?

**Comment by rusty1s** [comment]
Yes :)

**Comment by vymao** [comment]
Is this compatible with Dataset objects? I have a dataset of many Data objects in a Dataset corresponding to individual graphs that I currently process through the Dataloader. My hope was that I could use GraphSAINT on top of the individual graphs to sample them, as they are quite large.

**Comment by rusty1s** [comment]
It currently isn't. One thing that people do on multiple large graphs is to "batch" them into a single giant supergraph:
```python
from torch_geometric.data import Batch
data = Batch.from_data_list(dataset)
```
and use the resulting graph as input into `ClusterLoader`, ...
However, that's not ideal since it will sample subgraphs from different graphs into a mini-batch.
